pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,399 @@
+) {
+	controllerutil.SetControllerReference(derivedService, derivedEpSlice, r.Scheme())
+
+	derivedEpSlice.Labels = maps.Clone(derivedService.Labels)
+	if derivedEpSlice.Labels == nil {
+		derivedEpSlice.Labels = map[string]string{}
+	}
+	derivedEpSlice.Labels[mcsapiv1alpha1.LabelServiceName] = localEpSlice.Labels[discoveryv1.LabelServiceName]
+	derivedEpSlice.Labels[discoveryv1.LabelServiceName] = derivedService.Name
+	derivedEpSlice.Labels[mcsapiv1alpha1.LabelSourceCluster] = r.clusterName
+	derivedEpSlice.Labels[discoveryv1.LabelManagedBy] = endpointSliceLocalMCSAPIControllerName
+
+	if derivedEpSlice.Annotations == nil {
+		derivedEpSlice.Annotations = map[string]string{}
+	}
+	derivedEpSlice.Annotations[localEndpointSliceAnnotation] = localEpSlice.Name
+
+	derivedEpSlice.AddressType = localEpSlice.AddressType
+
+	// Beware those are shallow copies, content of the struct should not be modified
+	derivedEpSlice.Endpoints = slices.Clone(localEpSlice.Endpoints)
+	derivedEpSlice.Ports = slices.Clone(localEpSlice.Ports)
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) newDerivedEndpointSlice(
+	localEpSlice *discoveryv1.EndpointSlice, derivedService *corev1.Service,
+) *discoveryv1.EndpointSlice {
+	derivedEpSliceKey := getLocalDerivedEndpointSliceKey(localEpSlice)
+	if derivedEpSliceKey == nil {
+		return nil
+	}
+	derivedEndpointSlice := discoveryv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      derivedEpSliceKey.Name,
+			Namespace: derivedEpSliceKey.Namespace,
+		},
+	}
+
+	r.updateDerivedEndpointSlice(&derivedEndpointSlice, localEpSlice, derivedService)
+	return &derivedEndpointSlice
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var epSlice discoveryv1.EndpointSlice
+	if err := r.Client.Get(ctx, req.NamespacedName, &epSlice); err != nil {
+		return controllerruntime.Fail(client.IgnoreNotFound(err))
+	}
+
+	var localEpSlice *discoveryv1.EndpointSlice
+	var derivedEpSlice *discoveryv1.EndpointSlice
+	var derivedServiceName string
+	var err error
+
+	if epSlice.Labels[discoveryv1.LabelManagedBy] == endpointSliceLocalMCSAPIControllerName {
+		derivedEpSlice = &epSlice
+		localEpSlice, err = r.getLocalEndpointSlice(ctx, derivedEpSlice)
+		if err != nil {
+			return controllerruntime.Fail(err)
+		}
+		derivedServiceName = derivedEpSlice.Labels[discoveryv1.LabelServiceName]
+	} else {
+		localEpSlice = &epSlice
+		if isExported, err := r.isServiceExported(
+			ctx,
+			types.NamespacedName{
+				Name:      localEpSlice.Labels[discoveryv1.LabelServiceName],
+				Namespace: localEpSlice.Namespace,
+			},
+		); !isExported || err != nil {
+			return controllerruntime.Fail(err)
+		}
+		derivedEpSlice, err = r.getLocalDerivedEndpointSlice(ctx, localEpSlice)
+		if err != nil {
+			return controllerruntime.Fail(err)
+		}
+		derivedServiceName = getDerivedServiceName(localEpSlice)
+	}
+
+	if derivedServiceName == "" {
+		return controllerruntime.Success()
```

Comment:
Could be worth adding a comment on the reason for returning here, and maybe also emitting a log line just to simplify investigating possible unexpected behavior.

Reply:
Ok soooo I went ahead and tried to handle those as besides the user fiddling around the mirrored endpointslice which we could not handle (I still handled it though) there is actually a valid (but nice) case where the EndpointSlice is manually created and the user/something decide to remove the service label name instead of deleting it...